### DISCUSSION TITLE
Benchmarks

### DISCUSSION THREAD

**Discussion by msokalski** [artem-ogre/CDT#96] [Show and tell]
Hi!
Today I made benchmarks comparing **CDT**, **Delabella** (DLB) and **Delaunator** (DEL) performance.
(Yep, no Jonathan Richard Shewchuk's Triangle this time, sorry).

Interesting observations:
- CDT's duplicates removal / edge indices mapping speed drops significantly when number of points >= 1M.
- Erasing super-triangle (ES) is sometimes more expensive than erasing outer faces and holes (FF) in CDT.
- DLB's triangulation performance improves when shifting entire point cloud away from (0,0).
- DEL becomes unreliable and insanely slow on asymmetric distributions (i.e. gamma).

I'm also very curious how much DCT's performance would benefit from using lighter containers and directly linking faces with vertices and neighbor faces using pointers instead of indices.

Attached timing results and screenshots of input distributions. If you're interested in obtaining input files (in CDT's visualizer format), drop me a note.
[bench.zip](https://github.com/artem-ogre/CDT/files/9244522/bench.zip)

Here's zoomed in sample output (constraint edges are triangulation of input's random subset):
![bench](https://user-images.githubusercontent.com/14288504/182440121-199efbc3-7e49-4ffc-b100-712c7ba878c6.png)

**Comment by artem-ogre** [comment]
Neat and thanks for sharing! I’ll take a closer look when I’m back at the keyboard. Could you share the benchmarks code somehow?

**Reply by msokalski** [reply]
No problem, but it's ugly has heck.
https://github.com/msokalski/delabella/blob/IDelaBella3/delabella-sdl2.cpp#L2471
I didn't consider, anyone would look into it :) requires uncommenting bench_* things tho.

**Comment by msokalski** [comment]
Hi @artem-ogre !
There's option for randomized or 'as is goes' vertices insertion order in CDT.
Have you investigated other orderings that could theoretically speed up triangulation like in this paper:
https://www.researchgate.net/publication/257505494_A_new_insertion_sequence_for_incremental_Delaunay_triangulation
Any thoughts / conclusions?

**Reply by artem-ogre** [reply]
Got some time to read the paper, it's an interesting idea and should be quite straight-forward to implement. You mentioned you employed a different/opposite pre-sorting order. What is it?

**Reply by msokalski** [reply]
Right, in the paper they sort using breadth-first on Kd tree, currently I use depth-first sort on a slightly different tree. Kd requires perfect median cut while I use slicing bbox in half geometrically. So my tree don't have points in parent nodes, all are located in leafs only. I kept their idea of longer dimension cut instead of alternating scheme, indeed it works much better this way. Surprisingly, even if my tree is poorly balanced it allows triangulation to run faster, but probably for triangulation algorithm you use, their Kd breadth-first could be better.

**Reply by artem-ogre** [reply]
I see. Thanks a lot for the explanation!

**Reply by artem-ogre** [reply]
That's also how I do Kd-tree for finding the closest point: mid-split instead of median.

**Reply by msokalski** [reply]
My current insertion order "in action" :P
![delabella](https://user-images.githubusercontent.com/14288504/187897485-3dd3fcef-5185-4ea7-be16-25d5faa0595a.gif)

**Reply by artem-ogre** [reply]
I have implemented the bulk-loading scheme from the paper. It is indeed a very nice, elegant, and efficient idea! It is not just an insertion order: triangle-walk hints are included almost for free. Thank you very much for the paper recommendation.

The results are in this branch: https://github.com/artem-ogre/CDT/tree/new-insertion-sequence
New insertion order `VertexInsertionOrder::KdTreeBFS` is added.
I get 15-40% speed-ups which is really nice.

The only down-side of this method that the vertices can be bulk-loaded only once, with other schemes it was possible to arbitrarily chain multiple `insertVertices` and `insertEdges` calls. Currently I solved it by throwing and exception if `insertVertices` is called more than once with `VertexInsertionOrder::KdTreeBFS` order. Not very happy with it and am still thinking how to integrate all the features in a nice way. I would love to leave only one (best) insertion order but would not like to give up the ability to incrementally add vertices/edges.

**Reply by msokalski** [reply]
Sounds great, I'm definitely going to check it out, once I finish adding Shewchuck's Triangle to my benchmarking suite I'll re-run all tests once again. I can see the problem with multiple `insertVertices()` calls, currently I don't support it as well. I think having both initial bulk and incremental ongoing insertions would be something acceptable.

**Reply by msokalski** [reply]
Yep, it speeds up about 2x !
Have you decided yet if it's going to land in the master?

**Reply by artem-ogre** [reply]
@msokalski 
Yes, I am adding a bunch or smaller micro-optimizations and plan to merge it to the master.

**Reply by artem-ogre** [reply]
@msokalski 
I think I will not be adding any more optimizations. I will only be finalizing the API for the bulk-loading before merging.

A note for benchmarking: to squeeze the best performance one needs to compile CDT with Boost (`CDT_USE_BOOST`): I use `boost::container::flat_set` instead of `unordered_set` for marking visited triangles during walking as it is faster for that use-case.

Thanks a lot!

**Reply by msokalski** [reply]
I can barely see (5-10%) speed up after enabling CDT_USE_BOOST, is this how much I should expect? _(I'm on windows)_

**Reply by artem-ogre** [reply]
> I can barely see (5-10%) speed up after enabling CDT_USE_BOOST, is this how much I should expect? _(I'm on windows)_

Yes, that's about right. It's not a big thing as most of the improvements in 16bb64bff630bee1cea717c278469fae9aced8e2
But with all the small improvements added together it should be slightly faster (around ~10-15% maybe).

**Reply by msokalski** [reply]
Ok. I've updated my benchmarks with your changes and legendary Triangle!
https://github.com/msokalski/delabella/blob/master/bench/bench.md

**Reply by artem-ogre** [reply]
Thank you for the effort! Still not too happy though :) I will try to improve it further.

**Reply by msokalski** [reply]
Maybe it's not the fastest (... currently), but it's lovely clean and insensitive to all hard cases I could imagine.
I think it would be nice to switch to divide and conquer algo, which is also Kd based and has a great potential to run on MT in parallel.
I'd do that similarly to what Triangle has but with longer axis cut instead of alternating scheme. This is well described here: https://dl.acm.org/doi/pdf/10.1145/282918.282923 (page 110) but it uses tree along X only and some "exotic" topology structure :)

**Reply by msokalski** [reply]
Hey @artem-ogre
Something strange is going on on **Linux** with KdTreeBFS order. I've re-checked windows, it is fine.
To make thing more mysterious, it doesn't occur on uniform distribution.
Any idea what can cause this?

`KdTreeBFS` on Linux  
![gam2_kd](https://user-images.githubusercontent.com/14288504/188853480-a7b707e5-3ce4-49b4-94f9-54a633ee9468.svg)

---

`Randomized` on Linux  
![gam2_rnd](https://user-images.githubusercontent.com/14288504/188853674-808378d9-316d-41c6-a91a-8dfbf180fc47.svg)

EDIT: Adding input snippet for above results:
```
        uint64_t seed = 0x00000000E6F82B72ULL;
        std::mt19937_64 gen{seed};
        std::gamma_distribution<double> gam(0.1, 2.0);
        for (int i = 0; i < n; i++)
        {
          v[i].x = gam(gen) + 50;
          v[i].y = gam(gen) + 50;
        }
```
note: 50,50 shift makes a lot of vertex dups, so it requires pre-filtering.

EDIT 2:
Wow, I can fix it with CDT_USE_BOOST!

**Reply by artem-ogre** [reply]
@msokalski 
Hmm, so is enabling or disabling `CDT_USE_BOOST` does fix this? I can try to profile this case.

**Reply by artem-ogre** [reply]
@msokalski 
What compiler do you use on Linux?
It could be that implementation of `std::unordered_set` in standard library on Linux is quite terrible for the task at hand.

**Reply by msokalski** [reply]
- enabling CDT_USE_BOOST fixes it.
- gcc compiler

**Reply by artem-ogre** [reply]
Hi @msokalski! 👋

Sorry for distracting you again: I made another performance improvement. Instead of tracking a complete list of all triangles adjacent to a vertex I store only one adjacent triangle per vertex. To get other adjacent triangles pivoting around the vertex is used.
The changes are here: https://github.com/artem-ogre/CDT/tree/new-insertion-sequence

This required quite a big rewrite, which took some time. An interesting finding was that `std::nth_element` implementation is not portable across different standard libraries. So now I have to carry a copy of `libc++` implementation alongside CDT code 😮‍💨

I have also combined `Randomized` and `KDTreeBFS` orders into a single `Auto` order: it will do initial bulk-load with Kd-tree and then lazily put points into another Kd-tree for subsequent insertions. Implementation is kind of tricky but I hope it is easier to use the library this way.

Regarding the strange behavior of Kd-tree+gamma distribution, do you have pre-calculated input I can use by any chance? Preferably big enough for meaningful profiling. I am having problem generating big enough dataset because of all the duplicates :)

**Reply by artem-ogre** [reply]
> * enabling CDT_USE_BOOST fixes it.
> * gcc compiler

I tried running small dataset many times and I don't see any significant performance difference.

**Reply by msokalski** [reply]
No need to be sorry, I really enjoy watching the progress.
`Auto` - is a nice touch. 

Please find attached input file with removed vertex dups.
EDIT: OOPS cast to float, I'll resend it ....

**Reply by msokalski** [reply]
This should be ok
[dump.zip](https://github.com/artem-ogre/CDT/files/9558335/dump.zip)

**Reply by artem-ogre** [reply]
> This should be ok [dump.zip](https://github.com/artem-ogre/CDT/files/9558335/dump.zip)

Should I also do the edge insertion? I can't see a significant difference when just inserting vertices with and without Boost, unfortunately.

Here's my benchmark code:

    auto [vertices, ee] = readInputFromFile("dump.txt");
    CDT::RemoveDuplicatesAndRemapEdges(vertices, ee);

    auto t1 = high_resolution_clock::now();

    Triangulation triangulation(CDT::VertexInsertionOrder::Auto);
    triangulation.insertVertices(vertices);

    auto t2 = high_resolution_clock::now();

**Reply by artem-ogre** [reply]
Here are the flamegraphs: 
[flamegraphs.zip](https://github.com/artem-ogre/CDT/files/9558399/flamegraphs.zip)
(It takes about 800-900ms on my machine)

**Reply by msokalski** [reply]
No, no. Edges are just a bonus. 
Here's what I get with this input:
```
old boost KdTreeBFS
--------------
cdt removing dups... 403 ms
cdt triangulation... 1677 ms
cdt forcing edges... 951 ms
cdt copying... 194 ms
cdt erasing outer and holes... 1426 ms
cdt erasing super... 1124 ms

old std KdTreeBFS
--------------
cdt removing dups... 399 ms
cdt triangulation... 1828 ms
cdt forcing edges... 941 ms
cdt copying... 191 ms
cdt erasing outer and holes... 1426 ms
cdt erasing super... 1141 ms

new boost Auto
--------------
cdt removing dups... 403 ms
cdt triangulation... 1390 ms
delabella-sdl2: CDT/CDT/include/Triangulation.hpp:1901: bool CDT::Triangulation<T, TNearPointLocator>::hasEdge(CDT::VertInd, CDT::VertInd) const [with T = double; TNearPointLocator = CDT::LocatorKDTree<double, 32, 32, 32>; CDT::VertInd = unsigned int]: Assertion `iT != noNeighbor' failed.
Aborted (core dumped)

new std Auto
--------------
cdt removing dups... 401 ms
cdt triangulation... 1491 ms
delabella-sdl2: CDT/CDT/include/Triangulation.hpp:1901: bool CDT::Triangulation<T, TNearPointLocator>::hasEdge(CDT::VertInd, CDT::VertInd) const [with T = double; TNearPointLocator = CDT::LocatorKDTree<double, 32, 32, 32>; CDT::VertInd = unsigned int]: Assertion `iT != noNeighbor' failed.
Aborted (core dumped)
```

**Reply by msokalski** [reply]
ahh that is different `seed` than I previously benchmarked :/
i'll retry it with 0x00000000E6F82B72ULL ...

EDIT:
So, now I'm really confused, i don't experience such huge slowdown without boost as I posted previously. 
The only one thing that has changed is probably I've made a reboot.

**Reply by artem-ogre** [reply]
Looks like an assertion fails when inserting edges. Is edges input valid? 

Another thing, assertions should only trigger in debug builds (as far as I know).
Do you measure in release with optimizations enabled?

**Reply by msokalski** [reply]
@artem-ogre is new-insertion-sequence branch up to date? It aborts with an assert here :(

**Reply by artem-ogre** [reply]
@msokalski see the comment above :)

**Reply by msokalski** [reply]
Oh, I'm too slow sorry for not refreshing thread before posting.
I compile it with -O3 (no -g), AFAIK asserts are not excluded from release builds automatically.
Right it fails in insertEdges, but input is valid i think. In worst case there are edges with v1==v2 but no crossings.
EDIT: also, older commit (with KdTreeBFS) works fine.

**Reply by msokalski** [reply]
Hey @artem-ogre 
I've made a small input of 100 points and 13 edges causing insertEdges to abort. It can be easily verified with older CDT_visualizer that input is valid. I hope it can help you finding the problem.
[dump.txt](https://github.com/artem-ogre/CDT/files/9565395/dump.txt)

**Reply by artem-ogre** [reply]
Awesome! Thanks a lot! I will look into it.

**Reply by msokalski** [reply]
You're welcome. BTW, how do you produce such nice SVG flamegraphs?
Is this also possible on windows?

**Reply by artem-ogre** [reply]
> You're welcome. BTW, how do you produce such nice SVG flamegraphs? Is this also possible on windows?

I am using google perftools in combination with [Brendan Gregg's FlameGraph scripts](https://github.com/brendangregg/FlameGraph).
I also had success using it with perf. It should also work with other profiles and it looks like Visual Studio profiles are supported.
Let me know if you are interested in some particular details.

**Reply by msokalski** [reply]
Thank you

**Reply by artem-ogre** [reply]
Hi, @msokalski, I finally managed to do the final bug fixing and merged the new insertion sequence. Thank you for your suggestion and help! 
I also took the liberty to add you to the contributors list in README.

**Reply by msokalski** [reply]
Oh! Hi @artem-ogre .
I'm so glad you managed to finalize this part. 🎖 I've noticed latest activity on your project but I didn't expect you adding me to the contributors, thanks!
I should have some time for testing / benchmarking in second half of March, currently I'm quite busy with other boring things.

**Comment by artem-ogre** [comment]
Hi @msokalski,
No, I did not investigate other options and I was not aware of this paper. 

My gut feeling is that by itself it will likely not provide significant speedups on most of the real-world data. But it is definitely an interesting topic to investigate and benchmark.

**Comment by msokalski** [comment]
Hey @artem-ogre,
I've added benchmark charts to my repo including your CDT timings. If you're interested it's here: https://github.com/msokalski/delabella/blob/master/bench/bench.md
Interestingly, CDT performance seams to be almost insensitive to the input points, it runs as a tank :)

**Reply by artem-ogre** [reply]
@msokalski Thanks! :) It is great to have snippets to reproduce inputs too.
As for performance, it’s a pity that it seems to not be that fast (comparatively) but asymptotically things are looking good.

Maybe switching from vertex lists to something like half-edge and pre-calculating kd-tree can give a good boost. It would be great to get some time for profiling/optimization. What are your thoughts on constant factor difference? Is there something you do differently?

**Reply by msokalski** [reply]
I'm glad you like it. I've observed slightly better performance of CDT on linux, where heap management is a bit lighter than on windows so maybe, at least partially, it's matter of allocators used in std::vector or so, especially on small number of points (100-10K).

Kd style pre-sorting helped me to overcome poor performance in edge cases like for `Symmetry` or `Circle` (it was terrible with original algorithm pre-sorting with X^2+Y^2 comparator) but It didn't speed up anything in general, actually it slowed down initial phase significantly. 

Main difference I can see is probably the locator, I don't have it :) If I'm right, you use it for locating already inserted point nearby currently being inserted one? In contrast I lookup triangles from youngest to oldest assuming (requiring) vertices were pre-sorted nicely so with high probability latest triangle contains a point which is close to the current point being inserted. 

**Finally it's hard to tell if that constant factor would hold for other test cases, I've spent some time optimizing my code just on those tests and ended up with quite opposite pre-sorting order than in the paper I post you a link.**
